packages/firestore/src/remote/serializer.ts

```diff
@@ -972,34 +984,38 @@ export function toTarget(
   return result;
 }
 
-function toFilter(filters: Filter[]): ProtoFilter | undefined {
+function encodeFilters(filters: Filter[]): ProtoFilter | undefined {
   if (filters.length === 0) {
     return;
   }
-  const protos = filters.map(filter => {
-    debugAssert(
-      filter instanceof FieldFilter,
-      'Only FieldFilters are supported'
-    );
-    return toUnaryOrFieldFilter(filter);
-  });
-  if (protos.length === 1) {
-    return protos[0];
+
+  return encodeFilter(CompositeFilter.create(filters, CompositeOperator.AND));
+}
+
+function decodeFilters(filter: ProtoFilter): Filter[] {
+  const result = decodeFilter(filter);
+
+  // Instead of a singletonList containing AND(F1, F2, ...), we can return a list containing F1,
+  // F2, ...
+  // TODO(orquery): Once proper support for composite filters has been completed, we can remove
+  // this flattening from here.
+  if (
+    result instanceof CompositeFilter &&
+    compositeFilterIsFlatConjunction(result)
+  ) {
+    return Object.assign([], result.getFilters());
```

Comment:
The return type of `result.getFilters()` is `readonly []`, but the return type of this function is not `readonly`. So this line is copying the array members to a mutable array, to avoid propagation of the `readonly` modifier through the call stack where it would not really be appropriate.

I added a comment to clarify the intent of this line.

Reply:
👍  Thanks